data_structures/binary_search_tree.ts

```diff
@@ -0,0 +1,199 @@
+        if (!currentNode.rightChild) {
+          throw new Error('No such element found.');
+        }
+
+        currentNode = currentNode.rightChild;
+      } else {
+        if (!currentNode.leftChild) {
+          throw new Error('No such element found.');
+        }
+
+        currentNode = currentNode.leftChild;
+      }
+    }
+
+    return currentNode;
+  }
+
+  /**
+   * Inserts the given data into the binary search tree.
+   *
+   * @param data The data to be stored in the binary search tree.
+   * @returns
+   */
+  insert(data: T): TreeNode<T> {
+    if (!this.rootNode) {
+      this.rootNode = new TreeNode<T>(data);
+      return this.rootNode;
+    }
+
+    let currentNode: TreeNode<T> = this.rootNode;
+    while (true) {
+      if (data > currentNode.data) {
+        if (currentNode.rightChild) {
+          currentNode = currentNode.rightChild;
+        } else {
+          currentNode.rightChild = new TreeNode<T>(data);
+          return currentNode.rightChild;
+        }
+      } else {
+        if (currentNode.leftChild) {
+          currentNode = currentNode.leftChild;
+        } else {
+          currentNode.leftChild = new TreeNode<T>(data);
+          return currentNode.leftChild;
+        }
+      }
+    }
+  }
+
+  /**
+   * Finds the minimum value of the binary search tree.
+   *
+   * @returns The minimum value of the binary search tree
+   */
+  findMin(): T {
+    if (!this.rootNode) {
+      throw new Error('No root node defined');
+    }
+
+    const traverse = (node: TreeNode<T>): T => {
+      return !node.leftChild ? node.data : traverse(node.leftChild);
+    };
+
+    return traverse(this.rootNode);
+  }
+
+  /**
+   * Finds the maximum value of the binary search tree.
+   *
+   * @returns The maximum value of the binary search tree
+   */
+  findMax(): T {
+    if (!this.rootNode) {
+      throw new Error('No root node defined');
+    }
+
+    const traverse = (node: TreeNode<T>): T => {
+      return !node.rightChild ? node.data : traverse(node.rightChild);
+    };
+
+    return traverse(this.rootNode);
+  }
+
+  /**
+   * Traverses to the binary search tree in in-order, i. e. it follow the schema of:
+   * Left Node -> Root Node -> Right Node
+   *
+   * @param node The node to start from.
+   * @param array The already found node data for recursive access.
+   * @returns
+   */
+  inOrderTraversal(node?: TreeNode<T>, array: T[] = []): T[] {
+    if (node) {
+      this.inOrderTraversal(node.leftChild, array);
+      array.push(node.data);
+      this.inOrderTraversal(node.rightChild, array);
+    }
+
+    return array;
+  }
+
+  /**
+   * Traverses to the binary search tree in pre-order, i. e. it follow the schema of:
+   * Root Node -> Left Node -> Right Node
+   *
+   * @param node The node to start from.
+   * @param array The already found node data for recursive access.
+   * @returns
+   */
+  preOrderTraversal(node?: TreeNode<T>, array: T[] = []): T[] {
+    if (node) {
```

Comment:
Matter of taste: I'd prefer an early return `if (!node) return;` instead.

Reply:
As I used the early returns also before, this was simply an issue and for the sake of consistency, I changed it to your suggestion.